i10 Script to set number of intended children for all Assets

bkiahstroud opened the pull request:
Ref 
- https://github.com/scientist-softserv/ams/issues/10

bkiahstroud committed:
create BackfillAssetValidationStatus service

bkiahstroud committed:
filter out Assets whose child count is already set

Also, DRY up logic that writes IDs to a file

bkiahstroud committed:
notes

bkiahstroud committed:
locate original data source object

bkiahstroud committed:
extract :intended_children_count from original BatchItem source data

bkiahstroud committed:
don't null AdminData value unless attr is present

Before, if an attribute's key and value were completely missing, it would null out the value. I don't believe this is the desired behavior.

I interpret the `!env.attributes[key].present?` to mean "if an attribute **is present** and its value is missing, null out the data".

However, in reality, what this actually does is "if an attribute is present and its value is missing OR the attribute is missing entirely, null out the data", which feels too destructive

bkiahstroud committed:
extract :intended_children_count from original Bulkrax::Entry source data (XML only)

bkiahstroud committed:
extract logic into smaller methods

bkiahstroud committed:
clarify variable names

bkiahstroud committed:
add missing asset param

bkiahstroud committed:
fix NoMethodError "id" in error logging

bkiahstroud committed:
suppress most logging to see script progress

bkiahstroud committed:
reset processed IDs file when starting a fresh run

orangewolf commented on app/services/ams/backfill_asset_validation_status.rb:
```diff
@@ -0,0 +1,164 @@
+      if attrs_for_actor['intended_children_count'].blank?
+        raise StandardError, "Unable to count intended children for Asset #{asset.id}"
+      end
+
```
if you havent loaded the asset here - you can create a background job to do the actor stack and save and never load the asset in this loop.

orangewolf replied:
sidebar: what if there was an activerecord object called "CleanupRun" that had id, item_id, item_type, job_name, status, error columns.

bkiahstroud committed:
avoid querying Fedora if we don't have to

Querying both Solr and ActiveRecord once each is significantly faster than querying Fedora once

bkiahstroud committed:
refactor queries for readability

bkiahstroud committed:
extract slow Asset update to background job

Both reading to and writing from Fedora is very slow. To help the script run quickly, extract this logic to a background job where the worker can handle churning through these and pace matters less

bkiahstroud committed:
failedBackfillAssetValidationStatusJobs log errors

bkiahstroud committed:
track failed IDs in dedicated file

bkiahstroud committed:
write ID to processed file before most errors can occur

Since failed IDs are being tracked in their own file, it would be confusing if an ID showed up in the failed file but not the processed one. "Processed" is not meant to indicate success, but rather that the script has started processing that ID

bkiahstroud committed:
add method to reprocess failed IDs

bkiahstroud commented on app/services/ams/backfill_asset_validation_status.rb:
```diff
@@ -0,0 +1,180 @@
+
+    def initialize
+      setup_working_directory
+      # TODO: replace with tagged logger
```
Non-blocking

jeremyf commented on app/actors/hyrax/actors/asset_actor.rb:
```diff
@@ -63,6 +63,7 @@ def set_admin_data_attributes(admin_data, env)
 
         def should_empty_admin_data_value?(key, admin_data, env)
           key != :bulkrax_importer_id &&
+          env.attributes.key?(key) &&
```
This is a bit confusing.  And thinking about the how we can short circuit calling admin data; I suspect that's slower.

```ruby
key != :bulkrax_importer_id &&
env.attributes.key?(key) &&
!env.attributes[key].present? &&
admin_data.send(key).present?
```

bkiahstroud replied:
> And thinking about the how we can short circuit calling admin data; I suspect that's slower.

@jeremyf can you explain what you mean here? What is slower?

jeremyf replied:
Sure.  `env.attributes.key?` is almost certainly a hash access.  Whereas `admin_data` is an object and we're sending it the `key` message.  What does that method do?  It could query all of Fedora.

Also, by having that `admin_data` sandwiched between two `env.attributes` methods there's a bit of a disjointed logical flow; namely, I'm going from one subject to another and then back to that first subject.  And with the `&&` that feels a bit much.

bkiahstroud committed:
simplify logic determining if a given AdminData attribute should be nulled out

Removed checking if the AdminData's attribute has a value since setting a blank value to the same blank value isn't a big deal

jeremyf commented on app/actors/hyrax/actors/asset_actor.rb:
```diff
@@ -53,19 +53,19 @@ def save_aapb_admin_data(env)
+          return false if %i[bulkrax_importer_id hyrax_batch_ingest_batch_id].include?(key)
+
+          # The presence of the key with a "blank" value indicates we're intentionally emptying the value
+          env.attributes.key?(key) && env.attributes[key].blank?
```
What happened to the `admin_data` check?

bkiahstroud replied:
It seemed unnecessary. It was saying: 

> Only null out the value if a value is present 

I see no danger in the following, so I removed it: 

```ruby
admin_data.my_attribute
=> nil

admin_data.my_attribute = nil

admin_data.my_attribute
=> nil
```

jeremyf replied:
Agreed
